server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -156,34 +166,97 @@ public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listene
         return true;
     }
 
+    /**
+     * Add a listener for refreshes, calling it immediately if the location is already visible. If this runs out of listener slots then it
+     * forces a refresh and calls the listener immediately as well. The checkpoint cannot be greater than the processed local checkpoint.
+     *
+     * @param checkpoint the seqNo checkpoint to listen for
+     * @param listener for the refresh.
+     * @return did we call the listener (true) or register the listener to call later (false)?
+     */
+    public boolean addOrNotify(long checkpoint, ActionListener<Void> listener) {
+        assert checkpoint >= SequenceNumbers.NO_OPS_PERFORMED;
+        if (checkpoint <= lastRefreshedCheckpoint) {
+            listener.onResponse(null);
+            return true;
+        }
+        long maxIssuedSequenceNumber = maxIssuedSeqNoSupplier.getAsLong();
+        if (checkpoint > maxIssuedSequenceNumber) {
```

Comment:
Note that passing global checkpoints back and forth still can lead to dirty reads where not all documents in this shard with a seq no below gcp are "committed". In particular, the shard which is being hit here might be stale (i.e. not part of in-sync set anymore), and the code would therefore need to check that `request.waitForCheckpoint() <= shard.globalCheckpoint()`, i.e. the waitForCheckpoint is part of the history that noone can disagree on (only exception to this I can see is if you revised part of the history below gcp, e.g. restoring older snapshot / force-allocate stale primary, i.e. where the history-uuid changes).

An interesting variant of this is engine reset (IndexShard.resetEngineToGlobalCheckpoint), where an engine that previously had a max seq no of 5 can be reset to have max seq no of 4 with no more indexing operations coming in, so safeguarding against `> maxIssuedSequenceNumber` won't always help to ensure that the listener is not indefinitely stuck waiting here.

Reply:
I added functionality to ensure the sequence number is less than the global checkpoint before executing query.